PyriteCore/src/world/Tools/SceneRenderTools.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include "CommonConstantBuffers.h"
+#include "utils/math.h"
+#include "display/texture.h"
+
+#include "scene/scene.h"
+#include "world/camera.h"
+
+#include "display/RenderGraph/RenderGraph.h"
+#include "display/RenderGraph/BuiltinPasses/DepthPrePass.h"
+#include "display/RenderProfiles.h"
+
+namespace pyr
+{
+
+
+class SceneRenderTools
+{
+private:
+	
+	SceneRenderTools() = delete;
+
+	struct DepthOnlyDrawer
+	{
+		RenderGraph graph;
+		BuiltinPasses::DepthPrePass depthPass; // < note, works only for static meshes i think
+		std::shared_ptr<pyr::CameraBuffer>  pcameraBuffer = std::make_shared<pyr::CameraBuffer>();
+		DepthOnlyDrawer() : depthPass(512, 512)
+		{
+			graph.addPass(&depthPass);
+		}
+	};
+
+	struct DepthDrawer
+	{
+		pyr::Effect* depthOnlyEffect = nullptr;
+		pyr::Camera camera;
+		pyr::GraphicalResourceRegistry registry;
+		std::function<void(const RegisteredRenderableActorCollection& sceneDescription)> renderFn;
+		struct Buffers
+		{
+			std::shared_ptr<pyr::CameraBuffer>  pcameraBuffer = std::make_shared<pyr::CameraBuffer>();
+			std::shared_ptr<ActorBuffer>		pActorBuffer = std::make_shared<ActorBuffer>();
+		} buffers;
+
+		enum RenderType { Texture2D, TextureCube };
+		DepthDrawer(RenderType type)
+		{
+			std::vector<pyr::Effect::define_t> defines{};
+		
+			if (type == TextureCube) defines.push_back(pyr::Effect::define_t{ .name = "LINEARIZE_DEPTH", .value = "1"  });
+
+			depthOnlyEffect = registry.loadEffect(
+				L"res/shaders/depthOnly.fx",
+				InputLayout::MakeLayoutFromVertex<pyr::RawMeshData::mesh_vertex_t>(),
+				defines);
+
+			renderFn = [this](const RegisteredRenderableActorCollection& sceneDescription)
```

Comment:
why isn't that a member function?

Reply:
is it not ?

Reply:
it's a member variable of function type. it should be a plain function instead

Reply:
oh right i get it. I thought the function would be different depending on 3D / 2D, that's why i made it a lambda, but turns out it's no different. Cool